acra/acra-in-depth/data-structures/_index.md

```diff
@@ -2,3 +2,185 @@
+
+```AcraStruct = Begin_Tag + Throwaway_Public_Key + Encrypted_Random_Key + Data_Length + Encrypted_Data```
+
+- `Begin_Tag[8]` — 8 bytes, header tag (can be changed);
+- `Throwaway_Public_Key[45]` — temporary public key generated by AcraWriter;
+- `Encrypted_Random_Key[84]` — encrypted Random Key by using SMessage (see next);
+- `Data_Length[8]` — length of the Encrypted data (see next);
+- `Encrypted_Data[Data_Length]` — payload encrypted with Random Key.
+
+### Generating AcraStruct
+
+AcraWriter is used to generate AcraStruct, but the generation process is quite simple and can be implemented in any custom writer:
+
+- AcraWriter generates a keypair of throwaway keys using Themis EC key generator:<br/>
+  `Throwaway_Keypair = (Throwaway_Public_Key, Throwaway_Private_Key)`.
+- Generates Random Symmetric Key (`RK`), 32 bytes long.
+- Encrypts `RK` using [Secure Message](/pages/secure-message-cryptosystem/) with `Throwaway_Private_Key` and `Acra_Public_Key`(or Zone key – see [Zones](/pages/documentation-acra/#zones)):<br/>
+  `Encrypted_Random_Key = SMessage(RK, Throwaway_Private_Key, Acra_Public_Key)`.
+- Encrypts the payload with [Secure Cell](/pages/secure-cell-cryptosystem/) in Seal mode:<br/>
+  `Encrypted_Data = SCell(RK, payload)`.
+- Erases/fills with zeros memory area with the `RK`.
+- Calculates the encrypted payload length and transforms it into little endian 8 bytes long (`Data_Length`).
+- Connects attributes together as described in the original formula.
+- Erases/fills the memory area containing the `Throwaway_Keypair` and original payload with zeros.
+
+We recommend you to check out AcraStruct [examples](https://github.com/cossacklabs/acra/tree/master/examples) and try the [⚙️Acra Engineering Demo⚙️](https://github.com/cossacklabs/acra-engineering-demo/#what-is-this).
+
+### Decrypting AcraStruct
+
+AcraServer, upon receiving and detecting valid AcraStruct, is able to:
+
+- Extract Throwaway Public Key (`TPK`).
+- Decrypt asymmetric envelope with `TPK` and Acra's Private Key (or [Zone key](/pages/documentation-acra/#client-side-with-zones)).
+- Extract Random Key (`RK`) for Secure Cell container out of a decrypted envelope;
+- Decrypt Secure Cell, extract payload;
+- Reconstruct database answer in such a way that AcraStruct is replaced by decrypted data.
+
+Check implementation in [decryptor/base/utils.go](https://github.com/cossacklabs/acra/blob/master/decryptor/base/utils.go#L29).
+
+### Zone Ids
+
+Although not directly related to AcraStructs, Zone Ids are used for matching Zones to keys. The current format is 8 bytes "begin tag" + 16 symbols a-zA-Z.
+
+## Storage Models
+
+The storage model modes used in Acra are WholeCell and InjectedCell.
+
+In WholeCell mode, [AcraStruct](/pages/documentation-acra/#acrastruct) represents a complete piece of data (i.e. database cell, a file, or some data transmitted into [AcraTranslator](/pages/acratranslator/)). In this mode it is expected that the encrypted data will look something like:
```

Comment:
AcraServer can work with both variants of AcraStruct storage. When it stored inside other container or as whole piece of data. AcraTranslator works only with the whole AcraStruct.

Reply:
Should we may be explicitly state on this detail? For example, we can write:
```
There are two storage model modes used in AcraServer: WholeCell and InjectedCell (AcraTranslator supports only WholeCell mode)
```

and this would allow to avoid explaining this details in further text of this section. WDYT?